src/crimson/osd/ops_executer.cc

```diff
@@ -723,6 +724,114 @@ version_t OpsExecuter::get_last_user_version() const
   return pg->get_last_user_version();
 }
 
+void OpsExecuter::make_writeable(std::vector<pg_log_entry_t>& log_entries)
+{
+  const hobject_t& soid = obc->obs.oi.soid;
+  logger().debug("{} {} snapset={} snapc={}",
+                 __func__, soid,
+                 obc->ssc->snapset, snapc);
+
+  // clone?
+  if (head_existed &&                            // old obs.exists
+      snapc.snaps.size() &&                      // there are snaps
+      snapc.snaps[0] > obc->ssc->snapset.seq) {  // existing obj is old
+
+    //  clone object, the snap field is set to the seq of the SnapContext
+    //  at it's creation.
+    hobject_t coid = soid;
+    coid.snap = snapc.seq;
+
+  // existing snaps are stored desendingly in snapc,
+  // cloned_snaps vector will hold all the snaps stored until snapset.seq
+   std::vector<snapid_t> cloned_snaps = [&] {
+      auto last = std::find_if(
+        std::begin(snapc.snaps), std::end(snapc.snaps),
+        [&](snapid_t snap_id) { return snap_id <= obc->ssc->snapset.seq; });
+      return std::vector<snapid_t>{std::begin(snapc.snaps), last};
+    }();
+
+    // version
+    osd_op_params->at_version = pg->next_version();
+
+    auto snap_oi = prepare_clone(coid);
+
+    // make clone
+    do_write_op([this] (auto& backend, auto& os, auto& txn) {
+      return backend.clone(os, clone_obc->obs, txn);
+    }, true);
+    // TODO: maybe cache:
+    // encode(snap_oi, bv, get_osdmap()->get_features(CEPH_ENTITY_TYPE_OSD, nullptr));
+    // txn.setattr(coll->get_cid(), ghobject_t{obc->obs.oi.soid}, OI_ATTR, bv);
+    // txn.rmattr(coll->get_cid(), ghobject_t{obc->obs.oi.soid}, SS_ATTR);
+
+    delta_stats.num_objects++;
+    if (snap_oi->is_omap()) {
+      delta_stats.num_objects_omap++;
+    }
+    delta_stats.num_object_clones++;
+    // newsnapset is obc's ssc
+    obc->ssc->snapset.clones.push_back(coid.snap);
+    obc->ssc->snapset.clone_size[coid.snap] = obc->obs.oi.size;
+    obc->ssc->snapset.clone_snaps[coid.snap] = cloned_snaps;
+
+    // clone_overlap should contain an entry for each clone
+    // (an empty interval_set if there is no overlap)
+    if (obc->obs.oi.size) {
+      obc->ssc->snapset.clone_overlap[coid.snap].insert(0, obc->obs.oi.size);
+    }
```

Comment:
```
    auto &overlap = obc->ssc->snapset.clone_overlap[coid.snap];
    if (obc->obs.oi.size) {
      overlap.insert(0, obc->obs.oi.size);
    }
```
since you need to create an empty map at clone_overlap[coid.snap] unconditionally?

Reply:
The map will be created once the operator[] will be used. Why should we create it unconditionally?

Reply:
[aea0cb9be2b958dd9750add70c3b73ebbee64bcd](https://github.com/ceph/ceph/pull/46941/commits/aea0cb9be2b958dd9750add70c3b73ebbee64bcd#diff-04d5a3324b706d64a3230221d2eb08f6aa66829f93ca151ebb71896fb743ada7L8628) edited also here, will update accordingly.